### ISSUE TITLE
Improve Webservice Security: Add login flow and user scope permissions

### ISSUE THREAD

**Issue by elboletaire** [PrestaShop/PrestaShop#19291]
#### Problem

With the current webservice implementation, no one should directly use that for developing a frontend app directly connecting to prestashop webservice. Basically because there's only one way to authenticate to the api and, once authenticated, you can do whatever has been enabled in that api key. This brings us many problems:

- We don't have a proper API login method. Users can only use the API keys manually created in the control panel as a login method.
- Even though you could end up creating an api key for every user, this does not avoid users from accessing others users' data. If you gave the api key the permission to PUT on /orders, every user will be able to modify other users' orders.

The current Webservice implementation can bring security issues to a lot of users. I recently had to make a prestashop module doing exactly this*, because a developer built a frontend app for my client using a unique api key for the entire app, without taking in mind all the security issues that this brings.

> \* both a login method using users credentials, which creates an api key for that user with expiration and also scope protection, to avoid users trying to edit non-allowed entities.

#### Solution I'd expect

- Being able to use the webservice as a real in-production REST api. Current state is more like a middleware API, which needs an additional layer on top of it to avoid security issues.
- A proper login for users, using their credentials. Additionally, the field used for the login should be configurable (some websites want to use the e-mail, other's prefer other data like the NIF).
- Scope-protect all customer-related calls. Other users should not be able to alter non-owned entities.
- Being able to turn on/off these features from the webservice page (even though I do think these features must be enabled by default).
- Being able to configure the scope of the calls (?)

#### Considered alternatives

AFAIK, there are no alternatives for this.

**Comment by elboletaire** [comment]
BTW, this bot auto-comment is pretty useless.

**Comment by matks** [comment]
> BTW, this bot auto-comment is pretty useless.

This is a welcome message 😊useless but friendly !

ping @PrestaShop/prestashop-core-developers to answer this question/suggestion

**Comment by jolelievre** [comment]
Hi @elboletaire 

I agree with you that these WebServices should not be used in front production for all the reasons you mentioned. Or only for simple use cases like read operations. But for now the way it's implemented doesn't allow advanced permissions nor login feature.

I think it is more designed (as you said) as a middleware for ERP or any other third party software that would need to interact with the shop data.

Also the API have very few domain logic and validation (except field expected format basically) so it requires some knowledge on the modification made on the data to be sure it stays consistent. 

I'm not sure that plugging an authentication system per customer upon it would be enough to correctly deal with the features you mentioned. So it might require a huge work on our APIs which will probably be part of a totally new version of API (not a simple update/improvement of the existing Webservice API).

Tldr; This is definitely something that need and can be improved, but sadly I'm not sure when it can be done

**Comment by eternoendless** [comment]
This is an issue we're aware of. We'd like to build a front office API which basically would allow you to do everything the FO does right now but through an API. 

Right now it's not our priority to work on this project (our resources are limited), but we are in contact with some people who are interested in making this thing a reality. We'd like to collaborate and build it together with interested parties, but unfortunately we haven't made any real progress so far.

In the meantime I hear people rely on custom built APIs through modules.

**Comment by elboletaire** [comment]
> I'm not sure that plugging an authentication system per customer upon it would be enough to correctly deal with the features you mentioned. So it might require a huge work on our APIs which will probably be part of a totally new version of API (not a simple update/improvement of the existing Webservice API).

Agree. Plus there are some inconsistency issues in the current Webservice, plus everything is being properly integrated and using the symfony core; there are many reasons to revamp the entire Webservice, yes.

> Tldr; This is definitely something that need and can be improved, but sadly I'm not sure when it can be done

As said, I did something to (more or less) fix the security issues for my client, so it's not something I need, but is something I do think prestashop should have properly designed and implemented.

If I'm in this situation where someone (self-called "professional developer") used an api key giving ALL permissions to the publicly accessible front client... I'm pretty sure this won't be the only case.

In the meantime, at least an alert could be added to the current webservice configuration page, just trying to avoid people to wrongly use the webservice.

**Comment by jolelievre** [comment]
Hi again @elboletaire 

I would be curious about how you managed a safe auth for front, is it a sublayer of the current Webservice API? If you are willing to maybe this could be a feature to add through a PR. Maybe not something perfect like a total refacto, but at least an improvement for front uses?

**Comment by elboletaire** [comment]
I don't think this should be added to the core code because what I've done is adding an additional layer of security to the current webservice. Now my client app has two api keys defined in the webservice:

- The old one, with all the required permissions for the frontend app, only used internally by my module.
- The new one, only giving permissions to my "proper api" module (that's its name, yes..), used by the frontend app for everything.

All frontend requests need to have that api key with permissions to access my module, because my module does two things:

- Login. Checks user's credentials and, if valid, generates an api key for that user
- Proxy. Ensures the api key provided by the user (generated in the login) is valid and redirects all calls to the webservice using the other api key with all the actually valid permissions.

The proxy method also has an ordinary/vulgar/seedy scope check. I just basically forbid users from making some requests without specifying their own id_customer. If they try to alter or create an entity for another id_customer, it just throws an unauthorized exception. When doing this part I did realize that maybe, as said by you @jolelievre a few posts above, the api should be rewritten from scratch.

This works for my use case, and it's much more secure than the previous implementation, but I think we all will agree that this is not a valid solution for the prestashop core (for many reasons, but one of the most important is because it's a weird workflow, having to create multiple keys).

I do myself find a bit ugly my implementation, but couldn't think in a better way only via modules (because for me is really important to not touch Core files, even though I had to do it due to other issues) and with the given tools. The scoped limit is probably what's worst here, a truly scoped API should be implemented from scratch, IMHO.

Also, as I did create it as a module, its implementation into the prestashop core isn't possible *per se*. It would imply making changes to properly fit it.

So, I made this hoping it was gonna be a temporary solution. Directly adding this to the core would brake the implementation of, what I think should be, the real valid solution.

Finally, I've just started as freelancer and, even though I love doing stuff for the love of open source, I do need to eat, and I was considering in improving this module to add it to the prestashop modules store (if it's possible, ofc..).

**Comment by elboletaire** [comment]
Yesterday, while adding the information to the `EPIC` issue about the webservice, I realized that this issue was already created back in 2018: #11053. One or the other should be marked as duplicate (I don't care if it ends up being this one, although this contains more comments and information).

Anyway, don't forget to check the [epic issue about the webservice](https://github.com/PrestaShop/PrestaShop/issues/10069), where everything has been linked together.

**Comment by hibatallahAouadni** [comment]
Hello @elboletaire 

Thanks for the suggestion, the Product Team will take a look.
Please stay tuned !

Ping @PrestaShop/product-team , what do you think?

**Comment by elboletaire** [comment]
https://github.com/PrestaShop/PrestaShop/issues/9972#issuecomment-816610039
